src/riscv/lib/src/machine_state/memory/state.rs

```diff
@@ -359,12 +383,24 @@ where
     }
 }
 
+fn page_range(address: Address, length: NonZeroUsize) -> RangeInclusive<u64> {
+    let start_page = address_to_page_index(address) as u64;
+
+    let end_address = address
+        .wrapping_add(length.get() as Address)
+        .wrapping_sub(1);
```

Comment:
Should these not be saturating? Particularly relevant for the subtraction.

Reply:
no, it should be wrapping.

Particularly: 
- we know the length in non-zero. So the `wrapping_sub` will never underflow - unless the addition _already_ overflowed.
- take the case where we are writing 1 byte to `u64::MAX` (so the last byte of a page). The combination of wrapping add/wrapping sub ensures we remain 'within' the current page (we overflow to `0` and back to `u64::MAX`).
- if we write 2 bytes to `u64::MAX` however, we want to write to the next page too (the page starting at `0`). So we have to use the wrapping add. If we use saturating sub instead then the 1-byte case fails.